<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''SELECT
                COUNT(*) as vacancies
                FROM vacancies v
                '''    




In [5]:

df = pd.read_sql_query(query_3_1, connection) 
display(df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\2477442469.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


vacancies
0      49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
empl = f'''SELECT
           COUNT(*) as employers
           FROM employers e
           '''     

In [7]:
# результат запроса
empl_df = pd.read_sql_query(empl, connection)
display(empl_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\2801831759.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  empl_df = pd.read_sql_query(empl, connection)


employers
0      23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
count_areas = f'''SELECT
                  COUNT(*) as areas
                  FROM areas a 
                  '''

In [9]:
# результат запроса
areas_df = pd.read_sql_query(count_areas, connection)
display(areas_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\2608473996.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  areas_df = pd.read_sql_query(count_areas, connection)


areas
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
count_industries = f'''SELECT
                       COUNT(*) as industr_count
                       FROM industries i
                       ''' 

In [11]:
# результат запроса
industries_df = pd.read_sql_query(count_industries, connection)
display(industries_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\315854880.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  industries_df = pd.read_sql_query(count_industries, connection)


industr_count
0            294

***

### выводы по предварительному анализу данных
- Количество вакансий 49197.
- Количество работодателей 23501.
- Количество регионов 1362.
- Количество отраслей 294.

В среднем количество вакансий в два раза больше чем количество работодателей, что может указывать на повышенный спрос на рабочую силу (в среднем каждый работодатель нуждается в двух дополнительных работниках).

Количество регионов вызывает вопросы: что подразумивается под этим критерием? На данный момент в России насчитывается 89 различных региона, но никак не 1362. Данная категрия подлежит изучению: имеются ли повторения, все ли перечисленные регионы пренадлежат к России, подразумеваются под термином регионы города милионники и т.п.

Количество отрослей не вызывает вопросов. Можно сделать грубую крикидку средней нехватки работников в каждой отрасли: 49197/294=~167. Данная цифра ничего конкретного нам не говорит, кроме большой нехватки работников.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
top_vacancies_area = f'''
SELECT
    COUNT(v.id) as cnt,          -- Считает количество вакансий в каждой области (регионе) и присваивает этому значению псевдоним `cnt`.
    a.name as area               -- Извлекает название области (региона) и присваивает этому значению псевдоним `area`.
FROM 
    vacancies v                  -- Осуществляет выборку данных из таблицы `vacancies`, представляющей вакансии.
JOIN 
    areas a ON v.area_id = a.id  -- Выполняет соединение с таблицей `areas` по полю `area_id`, чтобы получить название региона, соответствующее каждой вакансии.
GROUP BY 
    a.name                       -- Группирует результаты по названию региона, чтобы получить количество вакансий для каждого региона.
ORDER BY 
    cnt DESC                     -- Сортирует результаты по количеству вакансий в убывающем порядке, чтобы на первом месте были регионы с наибольшим количеством вакансий.
LIMIT 5                          -- Ограничивает результат выборки пятью регионами, имеющими наибольшее количество вакансий.
'''


In [13]:
# результат запроса
top_vacancies_area_df = pd.read_sql_query(top_vacancies_area, connection)
display(top_vacancies_area_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\454989031.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  top_vacancies_area_df = pd.read_sql_query(top_vacancies_area, connection)


cnt             area
0  5333           Москва
1  2851  Санкт-Петербург
2  2112            Минск
3  2006      Новосибирск
4  1892           Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
vacancies_with_salary = f'''SELECT
                                COUNT(v.id)
                            FROM 
                                vacancies v
                            WHERE 
                                v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL
                            '''

In [15]:
# результат запроса
vacancies_with_salary_df = pd.read_sql_query(vacancies_with_salary, connection)
display(vacancies_with_salary_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\3298181002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_with_salary_df = pd.read_sql_query(vacancies_with_salary, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
average_salaries = f'''
                        SELECT
                            ROUND(AVG(v.salary_from)) as avg_salary_from,
                            ROUND(AVG(v.salary_to)) as avg_salary_to
                        FROM 
                            vacancies v
                        WHERE 
                            v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL
                    '''


In [17]:
# результат запроса
average_salaries_df = pd.read_sql_query(average_salaries, connection)
display(average_salaries_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\2434390514.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  average_salaries_df = pd.read_sql_query(average_salaries, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
type_sched_empl = f''' SELECT
                            v.schedule,
                            v.employment,
                            COUNT(v.id) as vacancy_cnt
                       FROM
                            vacancies v
                       GROUP BY 
                            v.schedule,
                            v.employment
                       ORDER BY
                            vacancy_cnt DESC
                       LIMIT 5    
                       '''

In [19]:
# результат запроса
type_sched_empl_df = pd.read_sql_query(type_sched_empl, connection)
display(type_sched_empl_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\3315839041.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  type_sched_empl_df = pd.read_sql_query(type_sched_empl, connection)


schedule           employment  vacancy_cnt
0       Полный день     Полная занятость        35367
1  Удаленная работа     Полная занятость         7802
2     Гибкий график     Полная занятость         1593
3  Удаленная работа  Частичная занятость         1312
4    Сменный график     Полная занятость          940

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса
needed_experience = f'''SELECT 
                            v.experience as need_exp,
                            COUNT(v.id) as cnt
                        FROM
                            vacancies v
                        GROUP BY
                            need_exp
                        ORDER BY
                            cnt
                        '''


In [21]:
# результат запроса
needed_experience_df = pd.read_sql_query(needed_experience, connection)
display(needed_experience_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\2634175964.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  needed_experience_df = pd.read_sql_query(needed_experience, connection)


need_exp    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

# выводы по детальному анализу вакансий
1. Топ-5 регионов по количеству вакансий. Ожидаемо в топе оказались крупные города такие как Москва, Санкт-Петербург, Новосибирск для России и Минск, Алматы для ближнего зарубежья. В таком результате нет ничего удивительного, так как крупные областные центры концентрируют в себе индустрию, крупный и малый бизнес, привлекают как работодателей, так и работников.
2. Примерно у половины всех вакансий заполнено хотябы одно из двух полей с зарплатой (salary_from, salary_to). Вероятно такие рабочие места будут привлекать больше желающих, так как на вакансиях с неуказанной заработной платой она скорее всего будет подбираться в индивидуальном порядке. Рабочие места с более четкими условиями труда являются более привлекательными для соискателей.
3. Среднии значения для "вилки" оплаты труда являются суммы 71065 и 110537 рублей что в обоих случаях является довольно высокой суммой. Основываясь на этом можно предположить, что работодателями указывающими в вакансиях размер оплаты труда скорее всего являются крупные работодатели, которые знаю что они хотят от работника и какую сумму готовы за это платить.
4. По количеству вакансий ледирует варинт "Полный день - Полная занятость" 35367 вакансий, с отставанием в 5 раз по количеству рабочих мест стоит "Удаленная работа - Полная занятость" 7802 вакансий, на третем месте с отставанием в 5 раз уже от второго места находится "Гибкий график - Полная занятость" 1593 вакансий. Как видно "Полная занятость" является лидером во всех вариантах графиков работы. Учитывая что общее количество представленных вакансий 49197, то вакансий с "Полной занятостью" порядка 90% из выборки, которые уже в себе делятся три категории (Полный день, Удаленная работа и Гибкий график).
5. Вакансий без опыта работы 7197 (около 15%), на такое количество мест могут расчитывать люди без опыта работы, студенты, только что отучившиеся и получившие работу. При построении модели стоит это учитывать. В основном требуется опыт работы до 3 лет и от 3 до 6 лет (около 80%). Опыт работы более 6 лет требуется только для 1337(менее 3%) рабочих мест(скорее всего это руководящие должности или места с повышенной ответственностью).

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса
employers_top =  f'''SELECT 
                            e.name as employer,
                            COUNT(v.id) as vac_cnt
                        FROM
                            vacancies v
                        JOIN
                            employers e ON v.employer_id = e.id    
                        GROUP BY
                            employer
                        ORDER BY
                            vac_cnt DESC
                        LIMIT 5    
                        '''


In [23]:
# результат запроса
employers_top_df = pd.read_sql_query(employers_top, connection)
display(employers_top_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\3258972006.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_top_df = pd.read_sql_query(employers_top, connection)


employer  vac_cnt
0         Яндекс     1933
1     Ростелеком      491
2       Тинькофф      444
3           СБЕР      428
4  Газпром нефть      331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [54]:
# текст запроса

top_empl_vac_area = f'''
SELECT        
    a.name,                      -- Извлекает название региона (области) и присваивает этому значению псевдоним `name`.
    COUNT(DISTINCT e.id) cnt_emp,  -- Считает количество уникальных работодателей в регионе, присваивая этому значению псевдоним `cnt_emp`.
    COUNT(DISTINCT v.id) cnt_vac  -- Считает количество уникальных вакансий в регионе, присваивая этому значению псевдоним `cnt_vac`.
FROM
    areas a
    LEFT JOIN employers e on a.id = e.area  -- Выполняет левое соединение с таблицей `employers`, чтобы получить работодателей, связанных с каждым регионом.
    LEFT JOIN vacancies v on v.area_id = a.id  -- Выполняет левое соединение с таблицей `vacancies`, чтобы получить вакансии, связанные с каждым регионом.
WHERE v.id is null               -- Фильтрует результаты, оставляя только те регионы, в которых нет вакансий (`v.id IS NULL`).
GROUP BY a.id                    -- Группирует результаты по идентификатору региона, чтобы получить количество работодателей и вакансий для каждого региона.
ORDER BY cnt_emp desc            -- Сортирует результаты по количеству работодателей в порядке убывания.
LIMIT 5                          -- Ограничивает результат пятью регионами с наибольшим количеством работодателей, но без вакансий.
'''


In [55]:
# результат запроса
top_empl_vac_area_df = pd.read_sql_query(top_empl_vac_area, connection)
display(top_empl_vac_area_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\1997128923.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  top_empl_vac_area_df = pd.read_sql_query(top_empl_vac_area, connection)


name  cnt_emp  cnt_vac
0              Россия      410        0
1           Казахстан      207        0
2  Московская область       75        0
3  Краснодарский край       19        0
4            Беларусь       18        0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# текст запроса
vacancies_in_area = f'''
SELECT
    COUNT(DISTINCT v.area_id) as areas_cnt,
    e.name as employer
FROM 
    employers e
LEFT JOIN 
    vacancies v ON v.employer_id = e.id
GROUP BY
    e.name
ORDER BY
    areas_cnt DESC

'''    

In [27]:
# результат запроса
vacancies_in_area_df = pd.read_sql_query(vacancies_in_area, connection)
display(vacancies_in_area_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\2987991428.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_in_area_df = pd.read_sql_query(vacancies_in_area, connection)


areas_cnt                employer
0            181                  Яндекс
1            152              Ростелеком
2            116              Спецремонт
3             88  Поляков Денис Иванович
4             71               ООО ЕФИН 
...          ...                     ...
23170          0                  СДЕЛКА
23171          0            Alandr Group
23172          0                     СДК
23173          0      Lemon Land Lombard
23174          0                  Махина

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
# текст запроса
no_indust_empl = f'''SELECT
    COUNT(DISTINCT e.id) AS cnt_employers
FROM
    employers e
LEFT JOIN
    employers_industries ei ON e.id = ei.employer_id
WHERE
    ei.industry_id IS NULL;

                        '''
    

In [29]:
# результат запроса
no_indust_empl_df = pd.read_sql_query(no_indust_empl, connection)
display(no_indust_empl_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\1719894510.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  no_indust_empl_df = pd.read_sql_query(no_indust_empl, connection)


cnt_employers
0           8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
# текст запроса
emp_name_four_ind = f'''SELECT 
                            e.name employer_name,
                            COUNT(industry_id) cnt
                        FROM 
                            employers e
                        JOIN
                            employers_industries i ON e.id = i.employer_id
                        GROUP BY 
                            employer_name
                        HAVING 
                            COUNT(i.industry_id) = 4    
                        ORDER BY 1
                        LIMIT 3
                        '''    
                           


In [31]:
# результат запроса
emp_name_four_ind_df = pd.read_sql_query(emp_name_four_ind, connection)
display(emp_name_four_ind_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\1941732459.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  emp_name_four_ind_df = pd.read_sql_query(emp_name_four_ind, connection)


employer_name  cnt
0  101 Интернет    4
1      21vek.by    4
2          2ГИС    4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# текст запроса
it_employers = f'''SELECT
                        COUNT(ei.employer_id) cnt
                    FROM
                        employers_industries ei
                    JOIN
                        industries e ON e.id = ei.industry_id
                    WHERE
                        e.name = 'Разработка программного обеспечения'
                    '''                            

In [33]:
# результат запроса
it_employers_df = pd.read_sql_query(it_employers, connection)
display(it_employers_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\3722539925.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  it_employers_df = pd.read_sql_query(it_employers, connection)


cnt
0  3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup

# URL страницы с городами-миллионниками
url = "https://ru.wikipedia.org/wiki/Города-миллионеры_России"

# Получаем HTML-код страницы
response = requests.get(url)
html = response.text

# Парсим HTML с помощью BeautifulSoup
soup = BeautifulSoup(html, 'html.parser')

# Находим таблицу с городами-миллионниками
table = soup.find('table', {'class': 'standard sortable'})

# Извлекаем все строки из таблицы, пропуская заголовок
rows = table.find_all('tr')[1:]

# Извлекаем названия городов из первой колонки таблицы
cities = [row.find_all('td')[1].text.strip() for row in rows]
# Преобразование списка городов в строку для SQL-запроса
cities_placeholder = ', '.join(f"'{city}'" for city in cities)

# Выводим результат
print(cities)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']


In [35]:
# текст запроса
yandex_vac = f'''
WITH temp AS (
    -- Создает временную таблицу `temp`, которая агрегирует данные по регионам.
    SELECT
        a.name AS region,        -- Извлекает название региона (области) и присваивает этому значению псевдоним `region`.
        COUNT(*) AS cnt          -- Считает количество вакансий в каждом регионе и присваивает этому значению псевдоним `cnt`.
    FROM 
        vacancies v              -- Осуществляет выборку данных из таблицы `vacancies`, представляющей вакансии.
    LEFT JOIN 
        areas a ON a.id = v.area_id  -- Выполняет левое соединение с таблицей `areas`, чтобы получить название региона для каждой вакансии.
    LEFT JOIN 
        employers e ON e.id = v.employer_id  -- Выполняет левое соединение с таблицей `employers`, чтобы получить работодателя для каждой вакансии.
    WHERE 
        e.name = 'Яндекс'        -- Фильтрует вакансии, оставляя только те, которые относятся к компании "Яндекс".
        AND a.name IN ({cities_placeholder})  -- Фильтрует вакансии по регионам, указанных в `cities_placeholder` (список регионов передается динамически).
    GROUP BY 
        a.name                   -- Группирует результаты по названию региона, чтобы получить количество вакансий для каждого региона.
    ORDER BY 
        cnt                      -- Сортирует результаты по количеству вакансий (в данной части `ORDER BY` влияет только на группировку в `temp`).
)
SELECT 
    region,                      -- Извлекает название региона из временной таблицы `temp`.
    cnt                          -- Извлекает количество вакансий для соответствующего региона.
FROM 
    temp
UNION ALL
SELECT 
    'Total',                     -- Добавляет строку с текстом "Total" (суммарное значение) в колонке региона.
    SUM(cnt)                     -- Считает общее количество вакансий для всех регионов, суммируя значения `cnt`.
FROM 
    temp
'''
               

In [36]:
# результат запроса
yandex_vac_df = pd.read_sql_query(yandex_vac, connection)
display(yandex_vac_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\1727464012.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  yandex_vac_df = pd.read_sql_query(yandex_vac, connection)


region    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4             Пермь   25.0
5            Казань   25.0
6    Ростов-на-Дону   25.0
7               Уфа   26.0
8            Самара   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            Total  485.0

# выводы по анализу работодателей
1. На первом месте расположился IT гигант Yandex с 1933 вакансиями, а на пятом месте Газпром Нефть с 331 вакансией. Если посмотреть на позиции расположенные между ними то можно заметить, что у Yandex вакансий больше, чем у следующих четырех позиций вместе взятых. Можно предположить что в IT-сфере имеется серьезный недостаток рабочей силы.
2. В топе регионов с работодателями и без вакансий расположились Россия	(410), Казахстан (207), Московская область	(75), Краснодарский край (19), Беларусь (18). В целом это говорит, что работадатели размещали вакансии в других регионах, но не об отсутствии рабочих мест в данных регионах. Все регионы являются крупными образованиями, которые сами по себе вмещают в себе другие регионы. Вероятнее всего при составлении вакансий работодатель указывал название региона более конкретно. Сфера деятельности компании может распространяться на весь указанный регион, но должность указанная в вакансии может быть связанна с деятельностью по всему региону и поэтому нет необходимости указывать это.
3. По количеству регионов с заявками лидерами являются IT гиганты Yandex и Ростелеком (181 и 152 соответственно). На третем месте по количеству регионов организация Спецремонт (116). Четвертое место занимает некто Поляков Денис Иванович (88), пятое ООО ЕФИН (71). Последние две позиции вызывают вопросы. Кто это и кого представляют?
4. Количество работодателей не указавших сферу деятельности составляет 8419, что составляет значительную часть(больше трети) от общего числа работодателей 23501. 
5. У многих компаний указаны несколько сфер деятельности. Обычно такими являются большие компании.
6. 3553 компании указали в своих сферах деятельности разработку програмного обеспечения, что является ~15% от общего числа работодателей.
7. IT гигант Yandex широко представлен по стране и ведет деятельность во всех городах-милионниках. Только в этих городах имеется порядка 500 вакансий.

***

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [37]:
# текст запроса
data_vac = f'''
SELECT 
    COUNT(*)
FROM
    vacancies v
WHERE
    LOWER(v.name) LIKE '%data%' OR LOWER(v.name) LIKE '%данн%'
'''        

In [38]:
# результат запроса
data_vac_df = pd.read_sql_query(data_vac, connection)
display(data_vac_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\2086182049.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data_vac_df = pd.read_sql_query(data_vac, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [39]:
# текст запроса
fits_to_ds = f'''
SELECT 
    COUNT(*) AS matching_vacancies
FROM
    vacancies v
WHERE
    (
        LOWER(v.name) LIKE '%data scientist%' OR
        LOWER(v.name) LIKE '%data science%' OR
        LOWER(v.name) LIKE '%исследователь данных%' OR
        (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%') OR
        LOWER(v.name) LIKE '%machine learning%' OR
        LOWER(v.name) LIKE '%машинн%обучен%'
    )
    AND
    (
        LOWER(v.name) LIKE '%junior%' OR
        v.experience = 'Нет опыта' OR
        v.schedule = 'Стажировка'
    );
'''

In [40]:
# результат запроса
fits_to_ds_df = pd.read_sql_query(fits_to_ds, connection)
display(fits_to_ds_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\3031284124.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  fits_to_ds_df = pd.read_sql_query(fits_to_ds, connection)


matching_vacancies
0                  51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [41]:
# текст запроса
ds_sql_postgres = f'''
SELECT 
    COUNT(*) AS matching_vacancies
FROM
    vacancies v
WHERE
    (
        LOWER(v.name) LIKE '%data scientist%' OR
        LOWER(v.name) LIKE '%data science%' OR
        LOWER(v.name) LIKE '%исследователь данных%' OR
        (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%') OR
        LOWER(v.name) LIKE '%machine learning%' OR
        LOWER(v.name) LIKE '%машинн%обучен%'
    )
    AND
    (
        LOWER(v.key_skills) LIKE '%sql%' OR
        LOWER(v.key_skills) LIKE '%postgres%'
    );
'''

In [42]:
# результат запроса
ds_sql_postgres_df = pd.read_sql_query(ds_sql_postgres, connection)
display(ds_sql_postgres_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\713992335.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_sql_postgres_df = pd.read_sql_query(ds_sql_postgres, connection)


matching_vacancies
0                 229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [43]:
# текст запроса
ds_python = f'''
SELECT 
    COUNT(*) AS matching_vacancies
FROM
    vacancies v
WHERE
    (
        LOWER(v.name) LIKE '%data scientist%' OR
        LOWER(v.name) LIKE '%data science%' OR
        LOWER(v.name) LIKE '%исследователь данных%' OR
        (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%') OR
        LOWER(v.name) LIKE '%machine learning%' OR
        LOWER(v.name) LIKE '%машинн%обучен%'
    )
    AND
    (
        LOWER(v.key_skills) LIKE '%python%'
    );
'''

In [44]:
# результат запроса
ds_python_df = pd.read_sql_query(ds_python, connection)
display(ds_python_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\3749165758.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_python_df = pd.read_sql_query(ds_python, connection)


matching_vacancies
0                 357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [45]:
# текст запроса
key_skills_ds_mean = f'''
SELECT 
    ROUND(AVG(LENGTH(v.key_skills) - LENGTH(REPLACE(v.key_skills, CHR(9), '')) + 1), 2)
FROM vacancies v    
WHERE 
    (LOWER(v.name) SIMILAR TO '(%data scientist%|%data science%|%исследователь данных%|%machine learning%|%машинн%обучен%)' or 
    (v.name LIKE '%ML%' and (v.name not ILIKE '%HTML%'))) and v.key_skills is not null '''

In [46]:
# результат запроса
key_skills_ds_mean_df = pd.read_sql_query(key_skills_ds_mean, connection)
display(key_skills_ds_mean_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\4095463142.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  key_skills_ds_mean_df = pd.read_sql_query(key_skills_ds_mean, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [47]:
# текст запроса
ds_salary = f'''
SELECT
    v.experience,
    ROUND(AVG(
        COALESCE(salary_from, salary_to) + COALESCE(salary_to, salary_from)
    ) / 2) AS avg_salary
FROM
    vacancies v
WHERE
    (
        LOWER(name) LIKE '%data scientist%' OR
        LOWER(name) LIKE '%data science%' OR
        LOWER(name) LIKE '%исследователь данных%' OR
        LOWER(name) LIKE '%machine learning%' OR
        LOWER(name) LIKE '%машинн%обучен%' OR
        (name LIKE '%ML%' and name NOT ILIKE '%HTML%')
    )
    AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
GROUP BY
    v.experience
HAVING
    v.experience = 'От 3 до 6 лет';

'''
    

In [48]:
# результат запроса
ds_salary_df = pd.read_sql_query(ds_salary, connection)
display(ds_salary_df)

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\2289103629.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_salary_df = pd.read_sql_query(ds_salary, connection)


experience  avg_salary
0  От 3 до 6 лет    243115.0

***

# выводы по предметному анализу
1. 1771 вакансий тем или иным образом имеют отношение к данным. (1771/49197) это составляет 3,5% от общего числа вакансий представленных в выборке.
2. Существует 51 различное наименование для вакансий связанных с обработкой данных.
3. Вакансий связанных с SQL или PostgreSQL насчитывается 229. Примерно 1/8 вакансий DS.
4. Знание Python тредуется на 357 вакансиях. Это примерно 1/5 вакансий по DS.
5. В среднем для DS вакансий указывают 6,4 навыка, интересно посмотреть весь список ключевых навыков.
6. Человек с опытом работы в DS от 3 до 6 лет может расчитвать на среднюю з/п в 243115 рубля. Мотивирует учиться дальше.


# Общий вывод по проекту

### подведем итог исследования, обобщите выводы
### здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

### Дополнительные исследования данных

In [49]:
# Посчитаем количество уникальных наименований вакансий в выборке

vacancies_list = f''' 
SELECT
    COUNT(DISTINCT(v.name))
FROM
    vacancies v
    
ORDER BY
    1;
'''

vacancies_list_df = pd.read_sql_query(vacancies_list, connection)
display(vacancies_list_df)    

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\217898114.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_list_df = pd.read_sql_query(vacancies_list, connection)


count
0  21223

Количество уникальных наименований представленных в выборке составляет 21223. Это немногим меньше половины всех вакансий. Это говорит нам о сильном разнообразии рынка труда. 

In [50]:
# Составим список и заодно псчитаем количество уникальных наименований для DS-вакансий

vacancies_list_ds = f''' 
SELECT DISTINCT
    v.name
FROM
    vacancies v
WHERE
    (
        LOWER(name) LIKE '%data scientist%' OR
        LOWER(name) LIKE '%data science%' OR
        LOWER(name) LIKE '%исследователь данных%' OR
        LOWER(name) LIKE '%machine learning%' OR
        LOWER(name) LIKE '%машинн%обучен%' OR
        (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
    )
ORDER BY
    v.name ASC;

'''

vacancies_list_ds_df = pd.read_sql_query(vacancies_list_ds, connection)
display(vacancies_list_ds_df) 

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\23862437.py:22: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_list_ds_df = pd.read_sql_query(vacancies_list_ds, connection)


name
0                                AI/ML Senior Engineer
1                               Analyst Data Scientist
2                               Analyst/Data Scientist
3    B2B Data Scientist (Эксперт по моделированию и...
4    Chief Data Science (Цифровой кредитный монитор...
..                                                 ...
281  Стажер направления Машинное обучение и анализ ...
282  Стажёр Data Scientist / Группа сервисов Антифр...
283                    Старший аналитик (Data Science)
284                  Старший разработчик (ML engineer)
285  Эксперт по машинному обучению и искусственному...

[286 rows x 1 columns]

Уникальными для вакансий DS 286 наименований из 1771. Полученую цифру нужно считать приблизительной поскольку написание названий вакансий в этом списке может отличаться всего в один знак.

In [51]:
# Посмотрим на список ключевых навыков для DS

key_skills_ds = f'''
WITH skill_list AS (
    SELECT 
        LOWER(UNNEST(STRING_TO_ARRAY(v.key_skills, CHR(9)))) AS skill
    FROM 
        vacancies v    
    WHERE 
        (LOWER(v.name) SIMILAR TO '(%data scientist%|%data science%|%исследователь данных%|%machine learning%|%машинн%обучен%)' OR 
        (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%')) 
        AND v.key_skills IS NOT NULL
)
SELECT DISTINCT
    skill
FROM 
    skill_list
ORDER BY 
    skill ASC;
'''

    
    
key_skills_ds_df = pd.read_sql_query(key_skills_ds, connection)
display(key_skills_ds_df)    

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\4024917935.py:24: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  key_skills_ds_df = pd.read_sql_query(key_skills_ds, connection)


skill
0                                  1c erp
1                         1c: бухгалтерия
2                              1с-битрикс
3                     1с: документооборот
4    1с: зарплата и управление персоналом
..                                    ...
482               управление требованиями
483           функциональное тестирование
484                               чат-бот
485                      широкий кругозор
486                  экономический анализ

[487 rows x 1 columns]

Всего для DS мы насчитали 487 различных уникальных ключевых навыков. Этот список подлежит дальнейшему изучению на выявления скрытых повторений, например когда навык пишется на разных языках, с ошибкой, или другими словами. 

In [52]:
# На какую зарплату может расчитывать начинающь специалист без опыта или с небольшим опытом

ds_salary_junior = f'''
SELECT
    v.experience,
    ROUND(AVG(
        COALESCE(salary_from, salary_to) + COALESCE(salary_to, salary_from)
    ) / 2) AS avg_salary
FROM
    vacancies v
WHERE
    (
        LOWER(name) LIKE '%data scientist%' OR
        LOWER(name) LIKE '%data science%' OR
        LOWER(name) LIKE '%исследователь данных%' OR
        LOWER(name) LIKE '%machine learning%' OR
        LOWER(name) LIKE '%машинн%обучен%' OR
        (name LIKE '%ML%' and name NOT ILIKE '%HTML%')
    )
    AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
GROUP BY
    v.experience
HAVING
    v.experience IN ('Нет опыта', 'От 1 года до 3 лет');

'''

ds_salary_junior_df = pd.read_sql_query(ds_salary_junior, connection)
display(ds_salary_junior_df)
	

C:\Users\monak\AppData\Local\Temp\ipykernel_16532\1718403341.py:28: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_salary_junior_df = pd.read_sql_query(ds_salary_junior, connection)


experience  avg_salary
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0

Как мы видим на этапе junior на большую зарплату можно не расчитывать. Давайте сравним со средней зарплатой в целом по выборке для вариантов "нет опыта" и "от 1 года до 3 лет".

In [53]:
salary_zero_exp= f'''
SELECT
    v.experience,
    ROUND(AVG(
        COALESCE(salary_from, salary_to) + COALESCE(salary_to, salary_from)
    ) / 2) AS avg_salary
FROM
    vacancies v
WHERE
    salary_from IS NOT NULL OR salary_to IS NOT NULL
GROUP BY
    v.experience
HAVING
    v.experience IN ('Нет опыта', 'От 1 года до 3 лет');

'''

salary_zero_exp_df = pd.read_sql_query(salary_zero_exp, connection)
display(salary_zero_exp_df)


C:\Users\monak\AppData\Local\Temp\ipykernel_16532\1868578974.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  salary_zero_exp_df = pd.read_sql_query(salary_zero_exp, connection)


experience  avg_salary
0           Нет опыта     40472.0
1  От 1 года до 3 лет     76541.0

Сравнение показывает, что в DS без опыта будешь получать столько же сколько с опытом "от 1 года до 3 лет" в среднем по выборке. Ощутимая разница.

# Общий вывод:

Исследование данных по вакансиям и работодателям выявило ряд интересных и важных тенденций на рынке труда, но также указало на необходимость дальнейшего углубленного анализа.

- Основные выводы:
1. Объем и разнообразие вакансий:

В выборке представлено 49,197 вакансий, охватывающих 23,501 работодателя, 1,362 региона и 294 отрасли. В среднем на одного работодателя приходится две вакансии, что может свидетельствовать о высоком спросе на рабочую силу.
Количество регионов в выборке значительно превышает официальное количество регионов в России, что требует уточнения. Возможно, в данных учитываются административные единицы, города или международные регионы, что могло привести к избыточной детализации.

2. Разнообразие наименований вакансий:
Количество уникальных наименований вакансий составляет 21,223, что составляет почти половину от всех вакансий. Это свидетельствует о большом разнообразии рынка труда и широком спектре предложений.
Для вакансий, связанных с Data Science (DS), количество уникальных наименований составило 286 из 1,771, что составляет примерно каждую шестую вакансию. Это число требует уточнения, так как незначительные различия в написании названий могут влиять на статистику.

3. Востребованность профессий DS:
Анализ показал, что профессии, связанные с DS, обработкой и анализом данных, имеют высокую востребованность. Средняя зарплата в этой области значительно выше, чем в среднем по выборке, что подтверждается соотношением уникальных вакансий к общему числу — 286/1,771 для DS и 21,223/49,197 для всех вакансий.
Для специалистов без опыта в DS предлагается заработная плата, сопоставимая с зарплатой для работников с опытом от 1 года до 3 лет по всем отраслям, что подчеркивает привлекательность этой сферы для начинающих специалистов.

4. Очистка и стандартизация данных:
Дальнейший анализ данных выявил наличие незначительных различий в наименованиях вакансий и ключевых навыков. Эти различия искажают статистику, и их влияние неизвестно. Необходима дополнительная работа по очистке данных, устранению дублирований и стандартизации записей.
Например, для вакансий DS выявлено 487 различных уникальных ключевых навыков, но этот список также нуждается в изучении и очистке, так как навыки могут быть записаны с ошибками, на разных языках или под разными именами.

- Рекомендации и варианты продолжения исследования:

1. Очистка и стандартизация данных:
Провести детальную очистку данных, включая исправление орфографических ошибок, устранение дублирований и унификацию написания ключевых навыков и названий вакансий. Это позволит получить более точные результаты и улучшить качество анализа.

2. Дополнительное исследование рынка DS:
Провести углубленный анализ рынка труда для профессий, связанных с DS, включая сравнение зарплат на разных уровнях опыта и по регионам. Это поможет лучше понять, какие навыки и опыт наиболее востребованы и как это влияет на уровень оплаты труда.

3. Анализ географического распределения:
Уточнить категорию «регионов» в данных, чтобы выяснить, какие административные единицы учитывались, и провести исследование по распределению вакансий между регионами. Это позволит выявить возможные диспропорции и региональные особенности рынка труда.

4. Анализ востребованности навыков:
Провести анализ востребованности отдельных ключевых навыков, особенно в области DS. Это поможет выявить, какие компетенции наиболее важны для работодателей и как их наличие влияет на предложение вакансий и уровень зарплат.

5. Сравнение с другими отраслями:
Сравнить уровень зарплат и требования по опыту работы в DS с другими высокотехнологичными отраслями, чтобы оценить конкурентоспособность и привлекательность этой сферы для новых специалистов.


Таким образом, исследование показало как востребованность профессий, связанных с анализом данных, так и необходимость дальнейшей работы с данными для получения более точных и полезных выводов. Очистка и стандартизация данных, а также углубленный анализ конкретных направлений, помогут лучше понять рынок труда и его потребности.






